x-pack/solutions/observability/plugins/streams/server/lib/streams/client.ts

```diff
@@ -0,0 +1,749 @@
+        return wiredDefinition;
+      })
+      .catch(async (error) => {
+        if (isElasticsearch404(error)) {
+          const dataStream = await this.getDataStream(name);
+          return await this.getDataStreamAsIngestStream(dataStream);
+        }
+        throw error;
+      })
+      .catch(async (error) => {
+        if (isElasticsearch404(error)) {
+          throw new DefinitionNotFound(`Cannot find stream ${name}`);
+        }
+        throw error;
+      });
+
+    return definition;
+  }
+
+  async getDataStream(name: string): Promise<IndicesDataStream> {
+    return this.dependencies.scopedClusterClient.asCurrentUser.indices
+      .getDataStream({ name })
+      .then((response) => {
+        const dataStream = response.data_streams[0];
+        return dataStream;
+      });
+  }
+
+  /**
+   * Creates an on-the-fly ingest stream definition
+   * from a concrete data stream.
+   */
+  private async getDataStreamAsIngestStream(
+    dataStream: IndicesDataStream
+  ): Promise<IngestStreamDefinition> {
+    const definition: IngestStreamDefinition = {
+      name: dataStream.name,
+      stream: {
+        ingest: {
+          routing: [],
+          processing: [],
+        },
+      },
+    };
+
+    definition.elasticsearch_assets = await getUnmanagedElasticsearchAssets({
+      dataStream,
+      scopedClusterClient: this.dependencies.scopedClusterClient,
+    });
+
+    return definition;
+  }
+
+  /**
+   * Checks whether the stream exists (and whether the
+   * user has access to it).
+   */
+  async existsStream(name: string): Promise<boolean> {
+    const exists = await this.getStream(name)
+      .then(() => true)
+      .catch((error) => {
+        if (isDefinitionNotFoundError(error)) {
+          return false;
+        }
+        throw error;
+      });
+
+    return exists;
+  }
+
+  /**
+   * Lists both managed and unmanaged streams
+   */
+  async listStreams(): Promise<StreamDefinition[]> {
+    const [managedStreams, unmanagedStreams] = await Promise.all([
+      this.getManagedStreams(),
+      this.getUnmanagedDataStreams(),
+    ]);
+
+    const allDefinitionsById = new Map<string, StreamDefinition>(
+      managedStreams.map((stream) => [stream.name, stream])
+    );
+
+    unmanagedStreams.forEach((stream) => {
+      if (!allDefinitionsById.get(stream.name)) {
+        allDefinitionsById.set(stream.name, stream);
+      }
+    });
+
+    return Array.from(allDefinitionsById.values());
+  }
+
+  /**
+   * Lists all unmanaged streams (ingest streams without a
+   * stored definition).
+   */
+  private async getUnmanagedDataStreams(): Promise<IngestStreamDefinition[]> {
+    const response =
+      await this.dependencies.scopedClusterClient.asCurrentUser.indices.getDataStream();
+
+    return (
+      response.data_streams
+        // if the template ends with @stream, it means we manage it
+        .filter((dataStream) => dataStream.template.endsWith('@stream') === false)
```

Comment:
nit: This check is a bit brittle and unnecessary since we prefer managed streams anyway in `listStreams`. We could just remove here.

Not a blocker